python/revng/tupletree/__init__.py

```diff
@@ -370,3 +389,323 @@ def init_check_kw_only(self, *args, **kwargs):
         original_init(self, *args, **kwargs)
 
     base_class.__init__ = init_check_kw_only  # type: ignore
+
+
+def construct_type(type_info: dict, data):
+    # MetaAddress comes already constructed so we jut return it
+    if type_info[type].__name__ == "MetaAddress":
+        return data
+
+    if type_info["ctor"] == "enum":
+        return type_info["possibleValues"](data)
+    elif type_info["ctor"] == "native":
+        return type_info[type](data)
+    if type_info["ctor"] == "parse":
+        if type_info[type].__name__ == "Type":
+            return type_info[type].from_dict(**data)
+        return type_info[type](data)
+    elif type_info["ctor"] == "class":
+        return type_info[type].from_dict(**data)
+
+
+class DiffSet:
+    def __init__(self, changes):
+        self.changes = changes
+
+    @classmethod
+    def from_dict(cls, **kwargs):
+        changes = []
+        for diff in kwargs["Changes"]:
+            changes.append(Diff(diff["Path"], diff.get("Add"), diff.get("Remove")))
+        return DiffSet(changes)
+
+    @classmethod
+    def yaml_constructor(cls, loader, node):
+        mapping = loader.construct_mapping(node, deep=True)
+        return cls.from_dict(**mapping)
+
+    @classmethod
+    def yaml_representer(cls, dumper: YamlDumper, instance) -> yaml.Node:
+        return dumper.represent_dict({"Changes": instance.changes})
+
+    def __repr__(self):
+        return repr({"Changes": self.changes})
+
+    def __str__(self):
+        return str({"Changes": self.changes})
+
+
+class Diff:
+    def __init__(self, path, add, remove):
+        self.Path = path
+        self.Add = add
+        self.Remove = remove
+
+    @classmethod
+    def yaml_representer(cls, dumper: YamlDumper, instance):
+        return dumper.represent_dict(instance._to_dict())
+
+    def _to_dict(self):
+        d = {"Path": self.Path}
+        if self.Add:
+            d["Add"] = self.Add
+        if self.Remove:
+            d["Remove"] = self.Remove
+        return d
+
+    def __repr__(self):
+        return repr(self._to_dict())
+
+    def __str__(self):
+        return str(self._to_dict())
+
+    def is_valid(self):
+        return self.Add is not None or self.Remove is not None
+
+
+YamlDumper.add_representer(Diff, Diff.yaml_representer)
+
+
+def _parse_diff(diffs: DiffSet, get_type_info: Callable) -> DiffSet:
+    new_changes = []
+    for diff in diffs.changes:
+        info = get_type_info(diff.Path)
+        if diff.Add and diff.Remove:
+            new_changes.append(
+                Diff(diff.Path, construct_type(info, diff.Add), construct_type(info, diff.Remove))
+            )
+        elif diff.Remove:
+            new_changes.append(Diff(diff.Path, None, construct_type(info, diff.Remove)))
+        elif diff.Add:
+            new_changes.append(Diff(diff.Path, construct_type(info, diff.Add), None))
+    return DiffSet(new_changes)
```

Comment:
The idea behind this is to parse the `yaml` into `objects` as soon as possible (when we load the diff) so that from that point on we only work with `objects`. Otherwise we need to construct the objects 2 times, in the `_validate_diff` and again in the `_apply_diff`. Beside that I think it makes sense that we deal with objects already constructed when we compare stuff.

Reply:
Is this approach OK? Can I close this conversation?

Reply:
I've had a think and I'm pretty sure this approach will work. The previous implementation was done before there were the `<Type>::` annotations in the diff, so it was impossible to determine the type without a model instance. 
I'll close this thread.